### PR TITLE
feat/pancakeswap v3 price calculation.

### PR DESCRIPTION
**Pull request by 0xcodercrane**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
- [x] Add an interface for v3 smart router  
- [x] Add the calculation method for the v3 smart router
- [x] Add the execution method for the v3 smart router

**Tests performed by the developer**:

**Tips for QA testing**:

### CODE REVIEW

**Comment by fengtality** [comment]
@vic-en can you take a look at this PR since you worked on PancakeSwap upgrade?

**Comment by vic-en** [comment]
@fengtality 
These changes are not related to pancake v3 LP.
They're changes to the existing swap.
Nevertheless, I'll look before today ends.

**Review by vic-en — COMMENTED** [review]
@0xcodercrane 
Just one comment for you, you can remove the router abi file as I see it's no longer used.

@fengtality 
As mentioned earlier, this changes pertain to the swap funtionalities.
As it is, it integrates the latest smart router which allow swaps to automatically use v3 pools when available.
That being said, there's no funtionality to strictly use v3 pool only as available in uniswap code. So I guess further decision is up to you.

**Comment by 0xcodercrane** [comment]
> @0xcodercrane Just one comment for you, you can remove the router abi file as I see it's no longer used.
> 
> @fengtality As mentioned earlier, this changes pertain to the swap funtionalities. As it is, it integrates the latest smart router which allow swaps to automatically use v3 pools when available. That being said, there's no funtionality to strictly use v3 pool only as available in uniswap code. So I guess further decision is up to you.

@vic-en  `routerAbi` is required field of `Uniswapish`.
and for the functionality which allow swaps to use only v3 pools, i pushed the changes about allowed pool types. i believe we can use that function.
https://github.com/hummingbot/gateway/pull/251/commits/45052d988e03acba7b1cd21c5ad91787b51a28ec

**Comment by cardosofede** [comment]
@vic-en @0xcodercrane @fengtality what about this pr, what else is missing to merge it?

**Comment by jellebuth** [comment]
We have been testing the implementation and noticed a 10x increase in RPC calls for price fetching. The free ANKR endpoint does not support price fetching for PCS V3 as there is a limit of 300/s and it is exceeded. 

Please take a careful look into this before merging. 

image

File: yarn.lock [hunk 1]
```diff
@@ -1123,137 +1123,137 @@
 "@ethersproject-xdc/abi@file:vendor/@ethersproject-xdc/abi":
   version "5.7.0"
   dependencies:
-    "@ethersproject-xdc/address" "file:vendor/@ethersproject-xdc/address"
```
Comment (fengtality) [inline yarn.lock:1126]: Please revert these changes to the `yarn.lock` file that change the XDC dependencies

**Review by fengtality — CHANGES_REQUESTED** [review]

File: yarn.lock (hunk 1 above)
Comment (0xcodercrane) [reply yarn.lock]: reverted

**Comment by 0xcodercrane** [comment]
Could you please approve ci/cd run @fengtality ?

**Comment by 0xcodercrane** [comment]
adding last fixes of unit test.

**Comment by cardosofede** [comment]
approved workflow run @0xcodercrane

**Comment by 0xcodercrane** [comment]
> approved workflow run @0xcodercrane

ty. how can I see which tests are failing from ci/cd. Would you help me on that? @cardosofede

**Comment by 0xcodercrane** [comment]
hey admins! anyone please approve the workflow run?

**Comment by fengtality** [comment]
@0xcodercrane Thanks - looks like tests are passing now. I'll review and address @vic-en's comment above shortly.

**Review by fengtality — CHANGES_REQUESTED** [review]
Looks good - thanks for the addition of the subgraph connection - I can definitely see why this improves latency for price fetching.

@jellebuth Thanks for testing this PR. Does the change from Ankr to LlamaRPC as default provider fix your issues?

@0xcodercrane Requested changes:

Review the fields in `pancakeswap.yml` in the templates folder. If any are not used anymore after this change, please remove them from this file.  This will let users know what options they have to modify how the connector works. 

@vic-en I don't think we need to add an option for users to select only V2 or V3 pools, unless they request it afterwards.

**Review by fengtality — APPROVED** [review]

**Comment by fengtality** [comment]
Actually, I will approve merge this PR in now, so that other outstanding PRs can update their dependencies. @0xcodercrane please clean up the Pancakeswap default configs template at your convenience.
